server.go

```diff
@@ -229,20 +232,93 @@ func newServer(listenAddrs []string, chanDB *channeldb.DB, cc *chainControl,
 	// CAN be passed into the ExternalIPs configuration option.
 	selfAddrs := make([]net.Addr, 0, len(cfg.ExternalIPs))
 	for _, ip := range cfg.ExternalIPs {
-		var addr string
-		_, _, err = net.SplitHostPort(ip)
+		ipLen := len(ip)
+		host, port, err := net.SplitHostPort(ip)
 		if err != nil {
-			addr = net.JoinHostPort(ip, strconv.Itoa(defaultPeerPort))
+			if (ipLen == 22 || ipLen == 62) && ip[ipLen-6:] == ".onion" {
+				// hidden service without a port
+				onionAddr := &torsvc.OnionAddress{
+					HiddenService: ip,
+					Port:          defaultOnionPort,
+				}
+				selfAddrs = append(selfAddrs, onionAddr)
+			} else {
+				// ipv4/6 address without a port
+				addr := net.JoinHostPort(ip, strconv.Itoa(defaultPeerPort))
+				lnAddr, err := cfg.net.ResolveTCPAddr("tcp", addr)
+				if err != nil {
+					return nil, err
+				}
+				selfAddrs = append(selfAddrs, lnAddr)
+			}
 		} else {
-			addr = ip
+			hostLen := len(host)
+			if (hostLen == 22 || hostLen == 62) && host[hostLen-6:] == ".onion" {
+				// hidden service with port
+				p, err := strconv.Atoi(port)
+				if err != nil {
+					return nil, err
+				}
+
+				onionAddr := &torsvc.OnionAddress{
+					HiddenService: host,
+					Port:          p,
+				}
+				selfAddrs = append(selfAddrs, onionAddr)
+			} else {
+				// ipv4/6 address with port
+				lnAddr, err := cfg.net.ResolveTCPAddr("tcp", ip)
+				if err != nil {
+					return nil, err
+				}
+				selfAddrs = append(selfAddrs, lnAddr)
+			}
 		}
+	}
+
+	// This code handles authenticating and dynamically creating v2 hidden
+	// services via Tor's ControlPort. The created v2 hidden service will
+	// be added to selfAddrs and used in the NodeAnnouncement message.
+	if cfg.torCtrl != nil {
+		// Set the server's TorControl struct. We will use the TorControl
+		// struct to dynamically create v2 hidden services.
+		s.torCtrl = cfg.torCtrl
 
-		lnAddr, err := cfg.net.ResolveTCPAddr("tcp", addr)
+		// Open the connection to Tor's ControlPort.
+		if err := s.torCtrl.Open(); err != nil {
+			return nil, err
+		}
+
+		// Authenticate to Tor's ControlPort.
+		if err := s.torCtrl.AuthWithPass(); err != nil {
+			return nil, err
+		}
+
+		// Create v2 hidden service via Tor's ControlPort.
+		hs, err := s.torCtrl.AddOnion()
```

Comment:
looks like this maybe can return the virtual port? Or maybe the complete `OnionAddress`?